### PR TITLE
Fix deadlock in test_pthread_dlopen_many

### PR DESCRIPTION
**Pull request by sbc100**
Avoid locking the DSO list during dlsync.  Instead only lock during dlopen (when adding new DSOs to the list).  This avoids a problem where static constructors were run during dlopen but other threads were blocked on dlsync while holding other locks (such as the malloc lock or the stdout lock).

### CODE REVIEW

File: system/lib/libc/dynlink.c [hunk 1]
```diff
@@ -50,15 +50,19 @@ static struct dso * _Atomic head, * _Atomic tail;
...
-static void dlsync_locked() {
+static void dlsync() {
   if (!thread_local_tail) {
     thread_local_tail = head;
   }
+  if (!thread_local_tail->next) {
+    return;
+  }
+  dbg("dlsync: catching up %p %p", thread_local_tail, tail);
   while (thread_local_tail->next) {
```
Comment (kripken) [inline system/lib/libc/dynlink.c:63]: Why is it ok to loop in this linked list without taking a lock? Can't it be modified by dlopen at the same time?

**Review by tlively — APPROVED** [review]
LGTM, although I admit I don't have the full context, so I will have missed any subtle bugs here 😬 

In general what is the strategy for ensuring concurrent reads and writes don't conflict? Is it the case that writes are atomic anyway, so reads can't observe inconsistent state? If so, why do we need a write lock?

File: system/lib/libc/dynlink.c (hunk 1 above)
Comment (sbc100) [reply system/lib/libc/dynlink.c:63]: This is a linked list that only grows and is never mutates, so it can be traversed in a lock-free way like this.

**Comment by sbc100** [comment]
> LGTM, although I admit I don't have the full context, so I will have missed any subtle bugs here grimacing
> 
> In general what is the strategy for ensuring concurrent reads and writes don't conflict? Is it the case that writes are atomic anyway, so reads can't observe inconsistent state? If so, why do we need a write lock?

Its important that each DSO is loaded in the right order.  If two threads begin the process of calling dlopen and race to add the next element that would be very bad.

I have several followup PRs that actually change the semantics here to keep the threads even more in lock step but I wanted to get this race fixed first since its occurring in CI.

**Pull request merged**